Account Keys Connection Status

lovo-h opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes #1806

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

This PR adds the ability to test whether a **Publishable Key** and **Secret Key** combination are valid. It does this by adding a "Test connection" link in the **Edit account keys & webhooks** modal (see image below). 

A tradeoff that we should be aware of, is that "test" tokens are created in the merchant's Stripe account when the connection is tested. These tokens do not contain any useful data but instead are simply used to test the validity of the publishable and secret keys.

![Screen Shot 2021-09-08 at 4 12 00 PM](https://user-images.githubusercontent.com/273592/132586292-52e6a387-b51e-43a2-80a1-5f4ddde61439.png)

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

- As a merchant, go to `WooCommerce > Settings > Payments > Stripe` and click on the `Settings` tab (at the top).
- Click on the `Edit account keys` button in order to open the **Edit test account keys & webhooks** modal.
- Enter publishable and secret keys and click on the `Test connection` link.
- Ensure publishable and secret keys result with a **Connection successful!** response, if they are valid. **We couldn't connect.**, otherwise.


---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

lovo-h committed:
Code cleanup: Remove repetitive code

Before, each store-module was spread into actions, selectors, and
resolvers. This created some repetition and would continue to create
more repetition in the future, as more store-modules would be added.
Now, the store-modules can be added into an array, which auto-handles
the spreading into actions, selectors, and resolvers.

lovo-h committed:
Code Cleanup: Capitalize the POST rest call

lovo-h committed:
Add frontend functionality to test account keys connection status

This adds the frontend UI that tests whether the publishable key and the
secret key are valid. It does this by creating a token with the
publishable key, using Stripe's library. Then retrieving the token with
the secret key, using our REST API, which uses Stripe's client library.

lovo-h committed:
Update tests to include newly added account-keys properties

Account key's store-module's state has newly added properties: isTesting
and isValid. It also includes newly added selectors. This commit
accounts for these new properties.

lovo-h committed:
Add frontend test to test account-keys connection status functionality

lovo-h committed:
Add backend functionality to test account keys connection status

This adds the backend functionality to determine if the secret key is
valid. It does this by retrieving a token with a given secret key and
token ID.

lovo-h committed:
Update changelog: Add ability to test Stripe account keys' validity

lovo-h committed:
Wrap account-key connection text on mobile

Before, the connection text would be too wide and would make the modal
footer's elements look scattered. This commit forces the connection
text's width to shrink, on mobile.

ricardo commented on client/settings/payment-settings/account-keys-modal.js:
```diff
@@ -183,12 +184,23 @@ export const AccountKeysModal = ( {
 
+	const onCloseHelper = () => {
+		// Reset AccountKeysConnectionStatus to default state.
+		updateIsValidAccountKeys( null );
```
I feel like we should reset to default state `onTabSelect` as well, so it doesn't seem like the connection test worked for both live and test keys.

And also because the form is reset when you switch tabs.

ricardo replied:
> And also because the form is reset when you switch tabs.

Although the "Save live/test keys" button is kinda doing that. Perhaps we should somehow let the user know about unsaved changes in a further issue?

lovo-h replied:
Good call on resetting the state to default `onTabSelect`.

As for letting the user know about the unsaved changes, I think [Alfredo recently attempted this](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2243#issuecomment-998333980) but was unsuccessful. But we could create a separate issue and try again. What do you think?

ricardo replied:
> But we could create a separate issue and try again. What do you think?

That'd be nice!

ricardo commented on client/settings/payment-settings/account-keys-connection-status.js:
```diff
@@ -0,0 +1,180 @@
+
+		try {
+			const stripe = await loadStripe( publishableKey );
+			const createTokenResult = await stripe.createToken( 'pii', {
+				personal_id_number: 'connection_test',
+			} );
```
I ran into a small issue while testing live keys with an HTTP URL:

> You may test your Stripe.js integration over HTTP. However, live Stripe.js integrations must use HTTPS.

Before looking at the console I didn't understand why it didn't work, had to check my keys a couple times.

Perhaps we should manually check for that error, or manually check if the website is using HTTPS, and offer a third message about having to test live keys with an HTTPS URL. What do you think?

lovo-h replied:
Good catch here. I think checking for that error is a good idea.

As for adding a third message, I think the given space is too limited. It may be better if we use a notification instead:

![image](https://user-images.githubusercontent.com/6844516/149174382-babf6554-2df6-498d-8ca7-d67b6a93ffe9.png)

That way, the given space is used specifically to indicate the status of the connection, and the notification is used to expand on details. What do you think?

ricardo replied:
That looks good to me! I'd just uppercase it to HTTPS.

ricardo commented on client/settings/payment-settings/account-keys-connection-status.js:
```diff
@@ -0,0 +1,180 @@
+			}
+
+			const tokenResult = await apiFetch( {
+				path: `/wc/v3/wc_stripe/tokens/${ tokenId }`,
```
Nit: Here we could use the `NAMESPACE` constant.

lovo-h replied:
Good call. I noticed a [separate example](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/client/settings/upe-toggle/provider.js#L42) where `NAMESPACE` was not used so I wasn't sure about the accepted approach.

I'll go ahead and update this and the other example 🙂

ricardo commented on client/data/account-keys/hooks.js:
```diff
@@ -43,6 +64,15 @@ export const useGetSavingError = () => {
 	}, [] );
 };
 
+export const useGetAccountKeysValidity = () => {
```
Is this being used somewhere?

lovo-h replied:
This is not being used. I'll go ahead and remove it 👍

ricardo commented on includes/admin/class-wc-rest-stripe-tokens-controller.php:
```diff
@@ -0,0 +1,64 @@
+		$secret_key = $request->get_header( 'X-WCStripe-Secret-Key' );
+
+		// Ensure secret_key is validated.
+		if ( ! preg_match( '/^\w+$/', $secret_key ) ) {
+			return new WP_REST_Response( [ 'result' => 'bad_request' ], 400 );
+		}
```
Nit: Do we need to validate the key or would the catch/error be enough? Just asking in case we can avoid repeating the bad request response.

lovo-h replied:
I agree with the idea of removing some repetition. But in this case, if a `secret_key` with invalid syntax is given, do we want to add overhead by pinging Stripe's server, or do we want to short-circuit the call?

I don't think it will be common that invalid syntax will be input. So, I'll go ahead and remove this validation and let the try/catch block handle it 🙂

ricardo replied:
> do we want to add overhead by pinging Stripe's server, or do we want to short-circuit the call?

I guess it would depend on how Stripe handles invalid keys, and if invalid keys with the wrong format are any different. I guess we could simply rely on Stripe's response, as it's the only source of truth.

ricardo commented on client/settings/payment-settings/account-keys-connection-status.js:
```diff
@@ -0,0 +1,180 @@
+		}
+
+		try {
+			const stripe = await loadStripe( publishableKey );
```
Although the form doesn't allow saving in these cases, I guess an edge case where people _could_ run into a false positive of successful connection would be trying to paste a test key into the live key input or vice-versa.

It's unlikely people will run into this, but it'd be straightforward to prevent it.

We could manually check if the keys contain `test` or `live` keywords and do some validation along with the aforementioned HTTP/HTTPS check.

lovo-h replied:
Good call. But as mentioned in a [previous comment](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2273#discussion_r783249873), we could display a notification when an account key is entered in the wrong tab (live/test):

![image](https://user-images.githubusercontent.com/6844516/149178188-df1bfecd-8d7e-4e62-85ad-7e14a6f738c4.png)

![image](https://user-images.githubusercontent.com/6844516/149178313-8396a1b5-29d2-4912-93de-45d820856f02.png)

What are your thoughts on this approach?

ricardo replied:
I agree the inline space is not enough, and this notification looks good to me.

Is the notification icon something we already use, or are you adding it now? I'm asking because it's different from the error icon on the connection status component.

lovo-h replied:
It is not something we already use and I am adding it now. It's an emoji:  ❌

I haven't seen the notification `icon` property used elsewhere, other than in [this notification](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/client/data/account/actions.js#L68).

Should I just remove the icon, in this case?

ricardo replied:
My only concern with the emoji is that it could be inconsistent with other notifications and that it looks different in other operating systems https://emojipedia.org/cross-mark/.

> Should I just remove the icon, in this case?

If we have other error messages without any icon, I'd vote for removing it, but if not, then we would possibly want a visual indication that this is an error. Then I'd ask Jarek or David about it.

ricardo reviewed (changes requested):
This feature is pretty cool 😎. Thanks for working on it!

PR tests well and the approach looks good to me.

> A tradeoff that we should be aware of, is that "test" tokens are created in the merchant's Stripe account when the connection is tested.

It's unfortunate that Stripe doesn't expose an API to directly validate the keys, but that's not a huge deal as the approach here looks clean and works well too.

ricardo commented on client/settings/payment-settings/account-keys-connection-status.js:
```diff
@@ -0,0 +1,180 @@
+				>
+					<DivSpinner />
+					<SpanConnectionText>
+						Testing connection...
```
Noticed this just now, but we should be localizing all these strings with `@wordpress/i18n`.

lovo-h replied:
Good call.

lovo-h committed:
Replace expanded REST API URL with NAMESPACE constant

Before, the REST API URL contained the expanded version as opposed to
using the NAMESPACE constant. This made it so that if the URL were to
ever be updated, it would require multiple places to be updated, as well.

lovo-h committed:
Remove unused code: useGetAccountKeysValidity

lovo-h committed:
Reset account-key validity upon switching between live/test tabs

Before, when account-keys were tested while on a certain tab and then
the tab was switched, the account-keys' status would remain. This would
be a bit confusing since the form is reset, once the user switches tabs.

lovo-h committed:
Internationalize text for account-key connection status

lovo-h committed:
Add notification for when incorrect account keys are entered

Before, test account-keys could be entered in the "live" tab and vice-
versa. This created inconsistent behavior.

lovo-h committed:
Handle live account-keys status connection on insecure (HTTP) connection

Live account-keys must use HTTPS connection. This commit handles the
case where an insecure (HTTP) connection is used to test a live account-
key.

lovo-h committed:
Allow Stripe to handle secret-key validation when testing keys

Removing backend validation to remove repetitive code and to allow
Stripe to validate the secret-key, instead.

ricardo commented on client/settings/payment-settings/account-keys-connection-status.js:
```diff
@@ -0,0 +1,232 @@
+			secretKey = keysToSave.secret_key;
+			if (
+				! publishableKey.startsWith( 'pk_live_' ) ||
+				! secretKey.startsWith( 'sk_live_' )
```
We must validate the use of restricted keys as well, starting with `rk_`, as mentioned in the input description.

lovo-h replied:
Good call. I didn't notice that.

ricardo commented on client/settings/payment-settings/account-keys-modal.js:
```diff
@@ -228,45 +240,59 @@ export const AccountKeysModal = ( {
 
 	const onTabSelect = ( tabName ) => {
+		// Reset AccountKeysConnectionStatus to default state.
+		updateIsValidAccountKeys( null );
```
One problem that prevents this from completely resetting it to its default state is that if there's still an ongoing request to validate the keys, the component can change after it has been reset 😕

We'd have to somehow cancel the request mid-flight in case it has been reset. I don't have a good suggestion on how to fix this right now, so feel free to ignore. I think it's fine as is 👍

lovo-h replied:
The same issue applies if the user closes the modal before the test is complete. 

One solution is to disable the **Cancel**, **Save test keys**, and remove the **close** button while the test is running:

![](https://cdn-std.droplr.net/files/acc_1209167/TVzVlP)

As for switching between tabs, what we can do is prevent the tab panel from switching if there's an ongoing connection test. This can be done by checking if `isTesting` validates to true. Unfortunately, we cannot prevent the tab panel tabs from thinking the tab body has in fact switched: 

![](https://cdn-std.droplr.net/files/acc_1209167/xpk7lC)

So, we may want to make this a separate issue and address it in the future. What are your thoughts on both of these solutions?

ricardo replied:
I'm personally not a fan of blocking the UI while the test is running 🤔. I think resetting the component would be best, but it's not a big deal if that can't be accomplished in this PR.

Ideally, I think we should have the live and test keys tab work independently, and allow you to save both keys at the same time. I believe that was the original design plan, as the title used to say "Edit account keys webhooks" and the button "Save keys".

For that, we would probably have two form views rendered at the same time, and one of them hidden according to the selected tab. Similarly, the connection status component could be duplicated for test/live and that problem would be fixed.

By duplicating it, I don't mean creating a totally separate component, but rendering the same component twice with a different prop.

What do you think about that?

> So, we may want to make this a separate issue and address it in the future.

I'd also vote for addressing this in a future issue, independently of the approach chosen.

lovo-h replied:
That sounds like a good approach. That would prevent the state from leaking from **Live** into **Test**, and vice-versa.

> I'd also vote for addressing this in a future issue, independently of the approach chosen.

Sounds good. I'll go ahead and make a separate issue 😎

lovo-h committed:
Allow connection status test for secret-keys that begin with 'rk'

Before, restrictive secret keys were incorrectly considered invalid
secret keys because they did not begin with the 'sk' prefix. Now, the
'rk' prefix is considered to be valid.

ricardo commented:
Not in the scope of this PR, but I just noticed the notification appears behind the modal in the DOM. This hides the notification in mobile devices and partially hides it in smaller screens.

Shall we create a separate issue?

<img width="726" alt="Notification behind modal" src="https://user-images.githubusercontent.com/5509901/149389644-3497a7b2-7d5a-4191-9206-05644091557c.png">

ricardo reviewed (approved):
LGTM 💯 🚢
